pkg/kubelet/secret_manager.go

```diff
@@ -0,0 +1,234 @@
+		if item.refCount == 0 {
+			delete(s.items, key)
+		}
+	}
+}
+
+func (s *secretStore) Get(namespace, name string) (*api.Secret, error) {
+	key := keyFunc(namespace, name)
+
+	s.Lock()
+	defer s.Unlock()
+	if item, ok := s.items[key]; ok {
+		return item.secret, nil
+	}
+	return nil, fmt.Errorf("secret no present")
+}
+
+func (s *secretStore) Refresh() {
+	s.Lock()
+	keys := make([]string, 0, len(s.items))
+	for key := range s.items {
+		keys = append(keys, key)
+	}
+	s.Unlock()
+
+	secrets := make([]*api.Secret, 0, len(keys))
+	for _, key := range keys {
+		namespace, name, err := namespaceNameFunc(key)
+		if err != nil {
+			glog.Errorf("Unexpected key transform error: %v", err)
+			secrets = append(secrets, nil)
+			continue
+		}
+		secret, err := s.kubeClient.Core().Secrets(namespace).Get(name)
+		if err != nil {
+			glog.Errorf("Unable to retrieve a secret %s/%s: %v", namespace, name, err)
+			secret = nil
+		}
+		secrets = append(secrets, secret)
+	}
+
+	s.Lock()
+	defer s.Unlock()
+	for i, key := range keys {
+		if secrets[i] == nil {
+			// Ignore updating to nil.
+		}
+		if item, ok := s.items[key]; ok {
+			item.secret = secrets[i]
+		}
+	}
+}
+
+// cachingSecretManager keeps a cache of all secrets necessary for registered pods.
+// It implements the following logic:
+// - whenever a pod is created or updated, the current versions of all its secrets
+//   are grabbed from apiserver and stored in local cache
+// - every GetSecret call is served from local cache
+// - every X seconds we are refreshing the local cache by grabbing current version
+//   of all registered secrets from apiserver
+type cachingSecretManager struct {
+	secretStore *secretStore
+
+	sync.Mutex
+	registeredPods map[string]*api.Pod
+}
+
+func newCachingSecretManager(kubeClient clientset.Interface) (secretManager, error) {
+	return &cachingSecretManager{
+		secretStore:    newSecretStore(kubeClient),
+		registeredPods: make(map[string]*api.Pod),
+	}, nil
+}
+
+func (c *cachingSecretManager) GetSecret(namespace, name string) (*api.Secret, error) {
+	return c.secretStore.Get(namespace, name)
+}
+
+func (c *cachingSecretManager) RegisterPod(pod *api.Pod) {
+	for _, reference := range pod.Spec.ImagePullSecrets {
+		c.secretStore.Add(pod.Namespace, reference.Name)
+	}
+	var prev *api.Pod
+	func() {
+		key := keyFunc(pod.Namespace, pod.Name)
+		c.Lock()
+		defer c.Unlock()
+		prev = c.registeredPods[key]
+		c.registeredPods[key] = pod
+	}()
+	if prev != nil {
+		for _, reference := range prev.Spec.ImagePullSecrets {
+			c.secretStore.Delete(prev.Namespace, reference.Name)
+		}
+	}
+}
+
+func (c *cachingSecretManager) UnregisterPod(pod *api.Pod) {
+	var prev *api.Pod
+	func() {
+		key := keyFunc(pod.Namespace, pod.Name)
+		c.Lock()
+		defer c.Unlock()
+		prev = c.registeredPods[key]
+		delete(c.registeredPods, key)
+	}()
+	if prev != nil {
+		for _, reference := range prev.Spec.ImagePullSecrets {
+			c.secretStore.Delete(prev.Namespace, reference.Name)
+		}
+	}
+}
```

Comment:
add a helper function to get all the referenced secret names from a pod (will be image pull secrets, secret volumes, and secrets in container/initContainer env valueFrom references), so we have that logic in one place